juju/juju
# Containers Maintain Their Own Link-Layer Data

## Description of change

This patch changes behaviour for setting container/KVM link-layer data.

Old:
- Containers have their link-layer data generated and written to state.
- This data is retrieved and used to provision the container.
- The data never changes and container machine agents sending link-layer updates is a no-op.

New:
- Containers now have link-layer data generated, but returned directly to the provisioner without writing to state.
- Containers are provisioned with the *same* interface configuration as they were previously.
- Container machine agents set and maintain their link-layer data.

One important change from this patch is that container devices will no longer reference their parents devices on the host. This change is already manifest on the 2.7 branch and an audit of the code indicates that this will not affect any Juju behaviour.

Various logic clean-ups accompany these changes, with redundant code removed where it is effectively a no-op, or is unused.

## QA steps

For at least LXD, AWS and MAAS (container networking methods local/fan/provider):
- Bootstrap, add a machine and add a LXD container to the machine.
- Check that once up, `juju show-machine 0` includes a container with link-layer device info.

For MAAS
- Add a KVM to the machine and check for link-layer devices when it comes up.

## Documentation changes

None.

## Bug reference

N/A
